File: ruby/difference-of-squares/difference_of_squares.rb
class Squares
  attr_reader :square_of_sum, :sum_of_squares, :difference
  
  def initialize(n)
    if n == 1
      @square_of_sum = 1
      @sum_of_squares = 1
      @difference = 0
    else
      @square_of_sum = ((1 + n) * n / 2.0) ** 2
      @sum_of_squares = n * (n + 1) * (2*n + 1) / 6
      @difference = @square_of_sum - @sum_of_squares
    end
  end
end

File: go/difference-of-squares/difference_of_squares.go
// Package diffsquares provides Difference of Squares solution.
package diffsquares

// SquareOfSum returns the square of the sum of the first n natural numbers.
func SquareOfSum(n int) int {
	sum := (n + 1) * n / 2
	return sum * sum
}

// SumOfSquares returns the sum of the squares of the first n natural numbers.
func SumOfSquares(n int) int {
	return n * (n + 1) * (2*n + 1) / 6
}

// Difference returns the difference between the square of the sum and the
// sum of the squares of the first n natural numbers.
func Difference(n int) int {
	return SquareOfSum(n) - SumOfSquares(n)
}
